File: 10.prime_and_composite_numbers/lesson_10_peaks.py
def solution(A):
    n = len(A)
    peaks = [n]*n
    peak_count = 0

    for i in range(n-2, 0, -1):
        if(A[i] > A[i-1] and A[i] > A[i+1]):
            peak_count += 1
            peaks[i] = i
        else:
            peaks[i] = peaks[i + 1]

    if(peak_count < 2):
        return peak_count
    
    peaks[0] = peaks[1]

    lenght_try = peak_count

    while(lenght_try != 0):
        lenght_try_curr = lenght_try
        if(n % lenght_try_curr == 0):
            i = 0
            sections = int(n / lenght_try_curr)
            while(i < n):
                #if we have a peak in the section
                if(peaks[i] < i + sections):
                    lenght_try_curr -= 1
                    i += sections
                    if(lenght_try_curr == 0):
                        return lenght_try
                else:
                    break
                
        lenght_try -= 1
                
    return 0

File: 10.prime_and_composite_numbers/lesson_10_peaks.js
function solution(A) {
    let n = A.length
    let peaks = new Array(n).fill(0)
    let peakCount = 0

    for(let i = 1; i < n - 1; i++) {
        if(A[i - 1] < A[i] && A[i] > A[i + 1]) {
            peaks[i] = 1
            peakCount ++
        }
    }
    if (peakCount == 0) {
        return 0
    }

    if (peakCount == 1) {
        return 1
    }

    let blocks = peakCount

    while(blocks > 0) {
        if(n % blocks == 0) {
            let blocksLength = n / blocks
            let blockEnd = blocksLength 
            let i = 0
            let foundFlag = false
            
            while(i < n) {
                foundFlag = false
                if(i < blockEnd) {
                    if(peaks[i] == 1) {
                        foundFlag = true
                        i = blockEnd
                        blockEnd += blocksLength
                    } else {
                        i++
                    }
                } else {
                    break
                }
            }    
            if(i >= n && foundFlag) {
                return blocks
            }     
        }
        blocks --
    }

    return blocks
}